oxide-auth/src/code_grant/extensions/pkce.rs

```diff
@@ -1,11 +1,46 @@
 use std::borrow::Cow;
+use std::error::Error;
+use std::fmt::{Display, Formatter};
 
-use crate::primitives::grant::{GrantExtension, Value};
+use crate::{
+    primitives::grant::{GrantExtension, Value},
+    util::avoid_alloc_cow_str_to_string,
+};
 
 use base64;
 use sha2::{Digest, Sha256};
 use subtle::ConstantTimeEq;
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PkceError {
```

Comment:
Even though I'm not yet certain about a contentful error type, surely not as an enum. It should be an opaque struct and this enum its `inner: ` attribute.